src/Symfony/Component/HttpKernel/Tests/Controller/ArgumentResolver/DateTimeValueResolverTest.php

```diff
@@ -0,0 +1,183 @@
+        $this->assertInstanceOf(FooDateTime::class, $results[0]);
+        $this->assertEquals('2016-09-08', $results[0]->format('Y-m-d'));
+    }
+
+    public function testDateTimeImmutable()
+    {
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTimeImmutable::class, false, false, null);
+        $request = $this->requestWithAttributes(['dummy' => '2016-09-08 00:00:00']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        $results = iterator_to_array($results);
+
+        $this->assertCount(1, $results);
+        $this->assertInstanceOf(\DateTimeImmutable::class, $results[0]);
+        $this->assertEquals('2016-09-08', $results[0]->format('Y-m-d'));
+    }
+
+    public function testInvalidDate404Exception()
+    {
+        $this->expectException(NotFoundHttpException::class);
+        $this->expectExceptionMessage('Invalid date given for parameter "dummy".');
+
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTime::class, false, false, null);
+        $request = $this->requestWithAttributes(['dummy' => 'Invalid DateTime Format']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        iterator_to_array($results);
+    }
+
+    public function testAttributeWithFormatInvalidDate404Exception()
+    {
+        $this->expectException(NotFoundHttpException::class);
+        $this->expectExceptionMessage('Invalid date given for parameter "dummy".');
+
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTime::class, false, false, null, false, [new MapDateTime(format: 'd.m.Y')]);
+        $request = $this->requestWithAttributes(['dummy' => '2012-07-21']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        iterator_to_array($results);
+    }
+
+    public function testAttributeWithYmdFormatInvalidDate404Exception()
+    {
+        $this->expectException(NotFoundHttpException::class);
+        $this->expectExceptionMessage('Invalid date given for parameter "dummy".');
+
+        $resolver = new DateTimeValueResolver();
+
+        $argument = new ArgumentMetadata('dummy', \DateTime::class, false, false, null, false, [new MapDateTime(format: 'Y-m-d')]);
+        $request = $this->requestWithAttributes(['dummy' => '2012-21-07']);
+
+        /** @var \Generator $results */
+        $results = $resolver->resolve($request, $argument);
+        iterator_to_array($results);
+    }
+
+    private function requestWithAttributes(array $attributes)
+    {
+        $request = Request::create('/');
+
+        foreach ($attributes as $name => $value) {
+            $request->attributes->set($name, $value);
+        }
+
+        return $request;
+    }
+}
+
+class FooDateTime extends \DateTime
```

Comment:
```suggestion
final class FooDateTime extends \DateTime
```
?

Reply:
nope, please let's not use final when not appropriate